buffer.c

```diff
@@ -3226,7 +3226,6 @@ evbuffer_add_file_segment(struct evbuffer *buf,
 			}
 		}
 	}
-	++seg->refcnt;
```

Comment:
Hm, but there is evbuffer_file_segment_free() in the error path, that will pre-decrement the refcnt, what I missed?

Reply:
BTW can you add a test? Looks like it is pretty easy to trigger by freezing the buffer

Reply:
Assuming we don't change the `evbuffer_add_file_segment()` function, there will always be a memory issue in `evbuffer_add_file()` regardless of whether the following line have been modified:
```
evbuffer_add_file(struct evbuffer *buf, int fd, ev_off_t offset, ev_off_t length)
{
      ......
	if (r == 0)
       ......
}
```

Reply:
Again, don't change the evbuffer_add_file_segment().

Restore line 3319 to 
```
if (r == -1)
```

```
buf = evbuffer_new()
evbuffer_add_file(buf, ) // refcnt = 2
evbuffer_free()   // seg will not be freed because refcnt = 1
```

```
buf = evbuffer_new()
evbuffer_freeze(buf, 0)
evbuffer_add_file(buf, ) // refcnt = 0
evbuffer_free() // double freed
```

Modify line 3319 to 
```
if (r == 0)
```

```
buf = evbuffer_new()
evbuffer_add_file(buf, ) // refcnt = 1
evbuffer_free()  // seg will be freed
```

```
buf = evbuffer_new()
evbuffer_freeze(buf, 0)
evbuffer_add_file(buf, ) // refcnt = 1
evbuffer_free() // seg will not be freed because buf can not find seg
```

Reply:
This PR can be covered by `test_evbuffer_freeze()` in `test\regress_buffer.c`

Reply:
Ok, so before this PR:

```
$ valgrind regress --timeout 0 --no-fork evbuffer/freeze_start
evbuffer/freeze_start: OK
1 tests ok.  (0 skipped)
==1415201== 
==1415201== HEAP SUMMARY:
==1415201==     in use at exit: 112 bytes in 2 blocks
==1415201==   total heap usage: 19 allocs, 17 frees, 13,984 bytes allocated
```

After:

```
$ valgrind regress --timeout 0 --no-fork evbuffer/freeze_start
evbuffer/freeze_start: OK
1 tests ok.  (0 skipped)
==1416141== 
==1416141== HEAP SUMMARY:
==1416141==     in use at exit: 0 bytes in 0 blocks
==1416141==   total heap usage: 19 allocs, 19 frees, 13,984 bytes allocated
```